src/app/main.rs

```diff
@@ -345,66 +431,128 @@ where
-                    .expect("inbound router");
+                // A per-`dst::Route` layer that uses profile data to configure
+                // a per-route layer.
+                //
+                // The `classify` module installs a `classify::Response`
+                // extension into each request so that all lower metrics
+                // implementations can use the route-specific configuration.
+                let dst_route_stack = phantom_data::layer()
+                    .push(metrics::layer::<_, classify::Response>(route_http_metrics))
+                    .push(classify::layer());
+
+                // A per-`DstAddr` stack that does the following:
+                //
+                // 1. Determines the profile of the destination and applies
+                //    per-route policy.
+                // 2. Annotates the request with the `DstAddr` so that
+                //    `RecognizeEndpoint` can use the value.
+                let dst_stack = endpoint_router
+                    .push(phantom_data::layer())
+                    .push(insert_target::layer())
+                    .push(buffer::layer())
+                    .push(profiles::router::layer(
+                        profile_suffixes,
+                        profiles_client,
+                        dst_route_stack,
+                    ));
+
+                // Routes requests to a `DstAddr`.
+                //
+                // 1. If the CANONICAL_DST_HEADER is set by the remote peer,
+                // this value is used to construct a DstAddr.
+                //
+                // 2. If the request is HTTP/2 and has an :authority, this value
+                // is used.
+                //
+                // 3. If the request is absolute-form HTTP/1, the URI's
+                // authority is used.
+                //
+                // 4. If the request has an HTTP/1 Host header, it is used.
+                //
+                // 5. Finally, if the Source had an SO_ORIGINAL_DST, this TCP
+                // address is used.
+                let dst_router = dst_stack
+                    .push(buffer::layer())
+                    .push(limit::layer(MAX_IN_FLIGHT))
+                    .push(router::layer(|req: &http::Request<_>| {
+                        let canonical = req
+                            .headers()
+                            .get(super::CANONICAL_DST_HEADER)
+                            .and_then(|dst| dst.to_str().ok())
+                            .and_then(|d| Addr::from_str(d).ok());
+                        info!("inbound canonical={:?}", canonical);
+
+                        let dst = canonical
+                            .or_else(|| super::http_request_authority_addr(req).ok())
+                            .or_else(|| super::http_request_host_addr(req).ok())
+                            .or_else(|| super::http_request_orig_dst_addr(req).ok());
```

Comment:
yeah, i had it all in one helper before but split it into separate helpers so that the logic is really really clear when you read the recognizer. I don't mind if it changes later but will leave this as-is.